[special] add `cupyx.scipy.special.log_softmax`

khushi-411 opened the pull request:
PR adds `cupyx.scipy.special.log_softmax`.

Benchmarks:

size | dtype | SciPy | CuPy
-----|-------|-------|-----
1000 | float32 | 0.041 ms | 0.244 ms
1000 | float64 | 0.049 ms | 0.296 ms
100000 | float32 | 0.592 ms | 0.432 ms
100000 | float64 | 1.420 ms | 0.312 ms
1000000 | float32 | 2.369 ms | 0.314 ms
1000000 | float64 | 12.358 ms | 0.942 ms


<details>
<summary>Script (Click Please)</summary>

<p>

```python
import cupy
import cupyx

import scipy

from cupy import testing
from cupyx.profiler import benchmark


n_warmup = 1
n_repeat = 10
dtype_set = ('float32', 'float64')
n_set = (1000, 100000, 1000000)


def get_time(pref):
    cpu_time = pref.cpu_times.mean()
    gpu_time = pref.gpu_times.mean()
    return max(cpu_time, gpu_time) * 1000  # ms


def time(a):
    cp_a = cupy.array(a)
    ref = scipy.special.log_softmax(a)
    ret = cupyx.scipy.special.log_softmax(cp_a)
    times = []
    perf = benchmark(scipy.special.log_softmax, (a,),
                     n_warmup=n_warmup, n_repeat=n_repeat)
    times.append(get_time(perf))
    perf = benchmark(cupyx.scipy.special.log_softmax, (cp_a,),
                     n_warmup=n_warmup, n_repeat=n_repeat)
    times.append(get_time(perf))
    return times


print('size | dtype | SciPy | CuPy')
print('-----|-------|-------|-----')
for n in n_set:
    for dtype in dtype_set:
        a = scipy.random.randint(0, 2, size=n).astype(dtype)
        times = time(a)
        print('{} | {} | {:.3f} ms | {:.3f} ms'.format(n, dtype, times[0], times[1]))
```

</p>
</details>

takagi commented on cupyx/scipy/special/_logsumexp.py:
```diff
@@ -0,0 +1,74 @@
+_log_softmax_kernel = cp._core.ReductionKernel(
+   'T x1',
+   'T y',
+   'exp(static_cast<float>(x1))',
+   'a + b',
+   'y = _post_map(a, &y)',
```
Let's use `exp` and `log` without casting to float to make things simple at first.

takagi replied:
```py
_log_softmax_kernel = cp._core.ReductionKernel(
   'T x1',
   'T y',
   'exp(x1)',
   'a + b',
   'y = log(a)',
   '0',
   name='log_softmax',
   preamble=_preamble
)
```

khushi-411 replied:
This was my first try. But I got the compilation error saying:
<details>
<summary>Error Log (Click Please)</summary>

<p>

```python
__________________________ TestLogSoftmax.test_log_softmax_large_inputs ____________________________
 
self = <cupyx_tests.scipy_tests.special_tests.test_log_softmax.TestLogSoftmax object at 0x7f36ad3341c0>
xp = <module 'cupy' from '/home/khushi/Documents/opensource/cupy/cupy/__init__.py'>
scp = <module 'cupyx.scipy' from '/home/khushi/Documents/opensource/cupy/cupyx/scipy/__init__.py'>
dtype = <class 'numpy.int8'>
 
    @testing.for_all_dtypes()
    @testing.numpy_cupy_allclose(scipy_name='scp', atol=atol, rtol=rtol)
    def test_log_softmax_large_inputs(self, xp, scp, dtype):
        a = xp.arange(4, dtype=dtype)
>       return scp.special.log_softmax(a)
 
tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cupyx/scipy/special/_logsumexp.py:71: in log_softmax
    out = _log_softmax_kernel(tmp, axis=axis, keepdims=True)
cupy/_core/_reduction.pyx:752: in cupy._core._reduction.ReductionKernel.__call__
    return self._call(
cupy/_core/_reduction.pyx:351: in cupy._core._reduction._AbstractReductionKernel._call
    cub_success = _cub_reduction._try_to_call_cub_reduction(
cupy/_core/_cub_reduction.pyx:700: in cupy._core._cub_reduction._try_to_call_cub_reduction
    _launch_cub(
cupy/_core/_cub_reduction.pyx:536: in cupy._core._cub_reduction._launch_cub
    _cub_two_pass_launch(
cupy/_core/_cub_reduction.pyx:471: in cupy._core._cub_reduction._cub_two_pass_launch
    func = _SimpleCubReductionKernel_get_cached_function(
cupy/_util.pyx:67: in cupy._util.memoize.decorator.ret
    result = f(*args, **kwargs)
cupy/_core/_cub_reduction.pyx:243: in cupy._core._cub_reduction._SimpleCubReductionKernel_get_cached_function
    return _create_cub_reduction_function(
cupy/_core/_cub_reduction.pyx:228: in cupy._core._cub_reduction._create_cub_reduction_function
    module = compile_with_cache(
cupy/_core/core.pyx:2229: in cupy._core.core.compile_with_cache
    return cuda.compiler._compile_module_with_cache(
cupy/cuda/compiler.py:492: in _compile_module_with_cache
    return _compile_with_cache_cuda(
cupy/cuda/compiler.py:570: in _compile_with_cache_cuda
    ptx, mapping = compile_using_nvrtc(
cupy/cuda/compiler.py:321: in compile_using_nvrtc
    return _compile(source, options, cu_path,
cupy/cuda/compiler.py:290: in _compile
    options, headers, include_names = _jitify_prep(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
 
source = '/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu\n#include <cupy/complex.cuh>\n#include <cupy/carra...ge\n  }\n\n  if (_tid == 0) {\n      type_mid_out& out0 = *(_out0 + blockIdx.x);\n      POST_MAP(aggregate);\n  }\n}\n'
options = ('-DFIRST_PASS=1', '--std=c++11', '-DCUPY_USE_JITIFY', '-I/home/khushi/Documents/opensource/cupy/cupy/_core/include', '-I/home/khushi/Documents/opensource/cupy/cupy/_core/include/cupy/_cuda/cuda-11', '-I/usr/local/cuda/include', ...)
cu_path = '/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu'
 
    def _jitify_prep(source, options, cu_path):
        # TODO(leofang): refactor this?
        global _jitify_header_source_map_populated
        if not _jitify_header_source_map_populated:
            from cupy._core import core
            _jitify_header_source_map = core._get_header_source_map()
            _jitify_header_source_map_populated = True
        else:
            # this is already cached at the C++ level, so don't pass in anything
            _jitify_header_source_map = None
    
        # jitify requires the 1st line to be the program name
        old_source = source
        source = cu_path + '\n' + source
    
        # Upon failure, in addition to throw an error Jitify also prints the log
        # to stdout. In principle we could intercept that by hijacking stdout's
        # file descriptor (tested locally), but the problem is pytest also does
        # the same thing internally, causing strange errors when running the tests.
        # As a result, we currently maintain Jitify's default behavior for easy
        # debugging, and wait for the upstream to address this issue
        # (NVIDIA/jitify#79).
    
        try:
            name, options, headers, include_names = jitify(
                source, options, _jitify_header_source_map)
        except Exception as e:  # C++ could throw all kinds of errors
            cex = CompileException(str(e), old_source, cu_path, options, 'jitify')
            dump = _get_bool_env_variable(
                'CUPY_DUMP_CUDA_SOURCE_ON_ERROR', False)
            if dump:
                cex.dump(sys.stderr)
>           raise JitifyException(str(cex))
E           AssertionError: Only cupy raises error
E           
E           Traceback (most recent call last):
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 255, in _jitify_prep
E               name, options, headers, include_names = jitify(
E             File "cupy/cuda/jitify.pyx", line 59, in cupy.cuda.jitify.jitify
E               cpdef jitify(str code, tuple opt, dict cached_sources=None):
E             File "cupy/cuda/jitify.pyx", line 92, in cupy.cuda.jitify.jitify
E               load_program(cuda_source, headers, nullptr, &include_paths,
E           RuntimeError: Runtime compilation failed
E           
E           During handling of the above exception, another exception occurred:
E           
E           Traceback (most recent call last):
E             File "/home/khushi/Documents/opensource/cupy/cupy/testing/_loops.py", line 45, in _call_func
E               result = impl(*args, **kw)
E             File "/home/khushi/Documents/opensource/cupy/tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py", line 23, in test_log_softmax_large_inputs
E               return scp.special.log_softmax(a)
E             File "/home/khushi/Documents/opensource/cupy/cupyx/scipy/special/_logsumexp.py", line 71, in log_softmax
E               out = _log_softmax_kernel(tmp, axis=axis, keepdims=True)
E             File "cupy/_core/_reduction.pyx", line 752, in cupy._core._reduction.ReductionKernel.__call__
E               return self._call(
E             File "cupy/_core/_reduction.pyx", line 351, in cupy._core._reduction._AbstractReductionKernel._call
E               cub_success = _cub_reduction._try_to_call_cub_reduction(
E             File "cupy/_core/_cub_reduction.pyx", line 700, in cupy._core._cub_reduction._try_to_call_cub_reduction
E               _launch_cub(
E             File "cupy/_core/_cub_reduction.pyx", line 536, in cupy._core._cub_reduction._launch_cub
E               _cub_two_pass_launch(
E             File "cupy/_core/_cub_reduction.pyx", line 471, in cupy._core._cub_reduction._cub_two_pass_launch
E               func = _SimpleCubReductionKernel_get_cached_function(
E             File "cupy/_util.pyx", line 67, in cupy._util.memoize.decorator.ret
E               result = f(*args, **kwargs)
E             File "cupy/_core/_cub_reduction.pyx", line 243, in cupy._core._cub_reduction._SimpleCubReductionKernel_get_cached_function
E               return _create_cub_reduction_function(
E             File "cupy/_core/_cub_reduction.pyx", line 228, in cupy._core._cub_reduction._create_cub_reduction_function
E               module = compile_with_cache(
E             File "cupy/_core/core.pyx", line 2229, in cupy._core.core.compile_with_cache
E               return cuda.compiler._compile_module_with_cache(
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 492, in _compile_module_with_cache
E               return _compile_with_cache_cuda(
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 570, in _compile_with_cache_cuda
E               ptx, mapping = compile_using_nvrtc(
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 321, in compile_using_nvrtc
E               return _compile(source, options, cu_path,
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 290, in _compile
E               options, headers, include_names = _jitify_prep(
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 263, in _jitify_prep
E               raise JitifyException(str(cex))
E           cupy.cuda.compiler.JitifyException: Runtime compilation failed
 
cupy/cuda/compiler.py:263: AssertionError
--------------------------------------- Captured stdout call ----------------------------------------
---------------------------------------------------
--- JIT compile log for /tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu ---
---------------------------------------------------
cupy/complex/complex.h(94): warning: __device__ annotation is ignored on a function("complex") that is explicitly defaulted on its first declaration
 
cupy/complex/complex.h(101): warning: __device__ annotation is ignored on a function("complex") that is explicitly defaulted on its first declaration
 
specializations/block_reduce_raking_commutative_only.cuh(98): warning: declaration does not declare anything
 
/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu(134): error: more than one instance of overloaded function "exp" matches the argument list:
            function "exp(double)"
__nv_nvrtc_builtin_header.h(149892): here
            function "exp(float)"
__nv_nvrtc_builtin_header.h(150793): here
            argument types are: (const type_mid_in)
 
/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu(127): warning: variable "_J" was declared but never referenced
 
1 error detected in the compilation of "/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu".
 
---------------------------------------------------
dtype is <class 'numpy.int8'>
```

</p>
</details>

takagi replied:
It's because T is integer type (`char` from numpy.int8) here, so the compiler couldn't find `exp(char)`. Would you begin to try suppling float32 and float64 values? `@for_dtypes('fd')` in test instead of `@for_all_dtypes()`. Also related to https://github.com/cupy/cupy/pull/6823#issuecomment-1170645216 .

khushi-411 replied:
Updated! Thanks :)

takagi commented on tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py:
```diff
@@ -0,0 +1,49 @@
+class TestLogSoftmax:
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_array_equal(scipy_name='scp')
```
Please use `numpy_cupy_allclose` to test floating point values.

takagi replied:
Supplying tolerance as following would work. (though, maybe you'll have more about dtypes)

```py
atol = {'default': 1e-6, cupy.float64: 1e-14}
rtol = {'default': 1e-6, cupy.float64: 1e-14}

@testing.for_all_dtypes()
@testing.numpy_cupy_allclose(scipy_name='scp', atol=atol, rtol=rtol)
def test_log_softmax_large_inputs(self, xp, scp, dtype):
    a = xp.arange(4, dtype=dtype)
    return scp.special.log_softmax(a)
```

khushi-411 replied:
Done

takagi commented:
What dtypes does `scipy.special.log_softmax` returns when it takes integral numbers? Plese cast inputs as such (I think using `ndarray.astype()` is enough).

takagi commented:
I checked e838fef, and it worked for float32/64 and complex64/128! Next, would you support integral dtypes as scipy does?

```py
>>> from scipy.special import log_softmax
>>> import numpy as np
>>> log_softmax(np.array([1], dtype=np.int8))
array([0.], dtype=float16)
>>> log_softmax(np.array([1], dtype=np.int16))
array([0.], dtype=float32)
>>> log_softmax(np.array([1], dtype=np.int32))
array([0.])
>>> log_softmax(np.array([1], dtype=np.int64))
array([0.])
```

takagi commented on cupyx/scipy/special/_logsumexp.py:
```diff
@@ -0,0 +1,50 @@
+import cupy as cp
+
+
+__all__ = ["log_softmax"]
```
nit: We conventionally use names that begin with an underscore for private things instead of `__all__` variable. Would you remove this line?

khushi-411 replied:
Thanks for sharing. I've removed that line :)

takagi commented:
Also, would you fix the committed file name `_logsumexp.py`?

takagi commented on tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py:
```diff
@@ -0,0 +1,49 @@
+@testing.with_requires('scipy')
+class TestLogSoftmax:
+
+    @testing.for_dtypes('fdFD')
```
Would you get these decorators back to test all dtypes? `@testing.for_all_dtypes(no_bool=True)`

Seems that SciPy's does not support boolean arrays for inputs, so `no_bool=True` option is set.

khushi-411 replied:
Fixed!

takagi commented:
/test mini

khushi-411 commented:
Hi, please wait. Some test cases didn't pass in my local system due to tolerance reasons.

takagi commented on tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py:
```diff
@@ -0,0 +1,49 @@
+import scipy.special  # NOQA
+
+
+atol = {'default': 1e-6, cupy.float64: 1e-14}
+rtol = {'default': 1e-6, cupy.float64: 1e-14}
```
You may add `cupy.float16: 1e-4` for float16, which generally has very low precision. (The value `1e-4` may vary. Please adjust it.)

khushi-411 replied:
Have updated the PR according to our offline discussions. Thanks a ton for your input and time :)

takagi commented:
/test mini

takagi commented:
Windows test says `error: identifier "thrust::FLT_MIN" is undefined in device code` . hmm...

takagi commented on cupy/_core/include/cupy/complex/arithmetic.h:
```diff
@@ -18,7 +18,7 @@
 
 #include <cupy/complex/complex.h>
-#include <cupy/complex/math_private.h>
+#include <cfloat>
```
Let me try including cfloat as the latest CUDA includes it in the corresponding thrust header instead of `math_private.h`.

khushi-411 replied:
Okay, thanks!

takagi commented:
/test mini

takagi commented:
/test mini

takagi commented:
/test mini

takagi commented:
Let us use 14847af for https://github.com/cupy/cupy/pull/6823#issuecomment-1172379994 as it makes the CI passes.

takagi commented on cupyx/scipy/special/_logsoftmax.py:
```diff
@@ -0,0 +1,64 @@
+
+    tmp = x - x_max
+
+    if tmp.dtype == cp.int8:
+        tmp = tmp.astype(cp.float16)
+    elif tmp.dtype == cp.int16:
+        tmp = tmp.astype(cp.float32)
+    elif tmp.dtype == cp.int32:
+        tmp = tmp.astype(cp.float64)
+    elif tmp.dtype == cp.int64:
+        tmp = tmp.astype(cp.float64)
+    elif tmp.dtype == cp.uint8:
+        tmp = tmp.astype(cp.float16)
+    elif tmp.dtype == cp.uint16:
+        tmp = tmp.astype(cp.float32)
+    elif tmp.dtype == cp.uint32:
+        tmp = tmp.astype(cp.float64)
+    elif tmp.dtype == cp.uint64:
+        tmp = tmp.astype(cp.float64)
```
Would you try the following works?
```py
if tmp.dtype.kind == 'i':
    for out_dtype in [cp.float16, cp.float32, cp.float64]:
        if cp.can_cast(tmp, out_dtype):
            tmp = tmp.astype(out_dtype)
            break
```

khushi-411 replied:
Hi! this doesn't work, we are getting same compilation error as before due to type casting.
<details>
<summary>Error Log (Click Please)</summary>

<p>

```python
raise CompileException(log, self.src, self.name, options,
cupy.cuda.compiler.CompileException: /tmp/tmpfknvsv0g/62e007c5affa579869ad99dfafae15f1f2ec15aa.cubin.cu(34): error: more than one instance of overloaded function "exp" matches the argument list:
            function "exp(double)"
__nv_nvrtc_builtin_header.h(149892): here
            function "exp(float)"
__nv_nvrtc_builtin_header.h(150793): here
            argument types are: (const T)

/tmp/tmpfknvsv0g/62e007c5affa579869ad99dfafae15f1f2ec15aa.cubin.cu(52): error: more than one instance of overloaded function "log" matches the argument list:
            function "log(double)"
__nv_nvrtc_builtin_header.h(149880): here
            function "log(float)"
__nv_nvrtc_builtin_header.h(150789): here
            argument types are: (_type_reduce)
```

</p>
</details>

takagi replied:
```
if tmp.dtype.kind in 'iu':
    for out_dtype in [cp.float16, cp.float32, cp.float64]:
        if cp.can_cast(tmp, out_dtype):
            tmp = tmp.astype(out_dtype)
            break
```

khushi-411 replied:
Works fine. Thanks

takagi reviewed (approved):
LGTM except for an additional comment!

takagi commented:
I got warnings on the tests. Would you adjust the test inputs?

```
=============================== warnings summary ===============================
tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py::TestLogSoftmax::test_log_softmax_ndarray_1
tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py::TestLogSoftmax::test_log_softmax_ndarray_2
tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py::TestLogSoftmax::test_log_softmax_ndarray_3
tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py::TestLogSoftmax::test_log_softmax_2d
tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py::TestLogSoftmax::test_log_softmax_3d
  /home/ext-mtakagi/.pyenv/versions/cuda-11.6/lib/python3.10/site-packages/scipy/special/_logsumexp.py:275: RuntimeWarning: overflow encountered in exp
    exp_tmp = np.exp(tmp)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
======================== 6 passed, 5 warnings in 17.37s ========================
```

khushi-411 commented:
Hi @takagi, I've updated the tests, but I see the following points:
- the tests work fine for `bhilefdFD`dtypes.
- overflows for `BHIL` dtypes and works fine only for input with a single value.

So, currently, I've separated the tests. Updated in [2ea00e3](https://github.com/cupy/cupy/pull/6823/commits/2ea00e3e65d377ae1f0c10db98ef1ec7d1abfbbc). Is that okay?

Thanks!

takagi commented:
Let me see the latest commit.

takagi commented on tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py:
```diff
@@ -0,0 +1,65 @@
+    @testing.for_dtypes('bhilefdFD')
+    @testing.numpy_cupy_allclose(scipy_name='scp', atol=atol, rtol=rtol)
+    def test_log_softmax_ndarray_1(self, xp, scp, dtype):
+        a = xp.arange(100, dtype=dtype)
```
Please properly adjust the input to avoid the overflow warning. I think `arange(1)` makes no sense as a test, and maybe `shaped_random` with properly scaled would work. (Also check the default value of `shaped_random`'s `scale` parameter)

takagi replied:
Even `np.exp(20)` makes inf in float16.
```
>>> np.exp(np.linspace(0, 100, num=10, dtype=np.int8))
array([1.000e+00, 5.987e+04,       inf,       inf,       inf,       inf,
             inf,       inf,       inf,       inf], dtype=float16)
```

khushi-411 replied:
> I think arange(1) makes no sense as a test, and maybe shaped_random with properly scaled would work.

Agreed with you. I've fixed it. it Is this okay now?

takagi replied:
> EDIT 2: Updated the tests. Is it okay to separate tests?

Let's skip unsigned integer tests in one test, leaving a comment as https://github.com/cupy/cupy/pull/6823#pullrequestreview-1036798376

takagi commented on tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py:
```diff
@@ -0,0 +1,83 @@
+@testing.with_requires('scipy')
+class TestLogSoftmax:
+
+    @testing.for_dtypes('bhilefdFD')
+    @testing.numpy_cupy_allclose(scipy_name='scp', atol=atol, rtol=rtol)
+    def test_log_softmax_ndarray_1(self, xp, scp, dtype):
+        a = testing.shaped_random((40, 50), xp, dtype=dtype)
+        return scp.special.log_softmax(a)
```
```suggestion
    @testing.for_all_dtypes(no_bool=True)
    @testing.numpy_cupy_allclose(scipy_name='scp', atol=atol, rtol=rtol)
    def test_log_softmax_ndarray_1(self, xp, scp, dtype):
        if xp.dtype(dtype).kind == 'u':
            # Unsigned integers make underflows in numpy (eventually seen as
            # overflows in `np.exp(tmp)`)
            pytest.skip()
        a = testing.shaped_random((40, 50), xp, dtype=dtype)
        return scp.special.log_softmax(a)```

takagi replied:
ah, I recalled that `x - x_max` underflows with unsigned integers, and eventually makes overflows in `np.exp(tmp)`. Let's skip tests for unsigned ints.

```
    tmp = x - x_max
    exp_tmp = np.exp(tmp)
```

khushi-411 replied:
Skipped the tests. Thanks for the inputs :)

takagi commented:
Thanks! I also checked locally.

takagi commented:
/test mini

takagi reviewed (approved):
LGTM!

khushi-411 commented:
Thanks @takagi for all the useful suggestions and for merging this PR!

takagi commented:
Would you also merge this PR, @kmaehashi ?

khushi-411 commented:
Thanks for merging, @kmaehashi!

leofang commented:
@takagi Do you know why we haven't hit the Windows issue until now? With the experiment in #6941 I see that this issue can also affect Windows builds, so I feel it's better to backport it.

takagi commented:
Honestly, I have no idea…

takagi commented:
please ignore https://github.com/cupy/cupy/pull/6823#issuecomment-1204643924
